ENH: Added max_gap keyword for series.interpolate

cchwala opened the pull request:
- [x] closes #12187 and #26796 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This PR introduces the new keyword `max_gap` for `interpolate`. For all NaN-gaps which are wider than `max_gap` no interpolation is carried out and all NaNs of the gap are preserved. This is in contrast to using the `limit` kwarg which does not prevent interpolating values in a longer NaN gap.

I added numpy-based implementation that searches for NaN-gaps wider than `max_gap`. In line with the current implementations for NaN handling in `series.interpolate`, a set of NaN-indices that has
to be preserved is generated. This is used in the end, after a full interpolation of all NaN is done, to restore the NaNs gaps that shall not be interpolated.

Test and documentation were also added.

It will need some small PEP8-cleanup and maybe tests using other interpolation methods then `linear` (edit: Done). But before I continue, I would like to get feedback if my approach is in general okay.

This PR might also be extended to close #16457 which is on interpolation directly after resampling.

Example usage:

```python
In [1]: import numpy as np                                                                                                                                                                                                         

In [2]: import pandas as pd                                                                                                                                                                                                        

In [3]: s = pd.Series([np.nan, 1., np.nan, 2., np.nan, np.nan,  
   ...:                5., np.nan, np.nan, np.nan, -1., np.nan, np.nan])                                                                                                                                                           

# Using the new `max_gap` kwarg
In [4]: s.interpolate(max_gap=2, limit_area='inside')                                                                                                                                                                               
Out[4]: 
0     NaN
1     1.0
2     1.5
3     2.0
4     3.0
5     4.0
6     5.0
7     NaN
8     NaN
9     NaN
10   -1.0
11    NaN
12    NaN
dtype: float64

# Compare to the result when using the existing `limit` kwarg
In [5]: s.interpolate(limit=2, limit_area='inside')                                                                                                                                                                                
Out[5]: 
0     NaN
1     1.0
2     1.5
3     2.0
4     3.0
5     4.0
6     5.0
7     3.5
8     2.0
9     NaN
10   -1.0
11    NaN
12    NaN
dtype: float64
```

Timing:
```python
In [6]: %timeit s.interpolate(max_gap=2, limit_area='inside')                                                                                                                                                                       
708 µs ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit s.interpolate(limit=2, limit_area='inside')                                                                                                                                                                        
631 µs ± 5.85 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
 
The relative speed difference is similar for larger Series.

cchwala committed:
Added maxgap keyword for series.interpolate

Added numpy-based implementation that searchs for NaN-gaps wider
than `maxgap`. In line with the current implementations for
NaN handling in `series.interpolate`, a set of NaN-indices that has
to be preserved is generated.

Test and documentation were also added.

pep8speaks commented:
Hello @cchwala! Thanks for submitting the PR.

 - In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/5e4b2ee4debadfbce3ed8ecbfb8f11f6ca05ee75/pandas/core/generic.py), following are the PEP8 issues :

> [Line 6652:75](https://github.com/pandas-dev/pandas/blob/5e4b2ee4debadfbce3ed8ecbfb8f11f6ca05ee75/pandas/core/generic.py#L6652): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 6653:70](https://github.com/pandas-dev/pandas/blob/5e4b2ee4debadfbce3ed8ecbfb8f11f6ca05ee75/pandas/core/generic.py#L6653): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 6655:1](https://github.com/pandas-dev/pandas/blob/5e4b2ee4debadfbce3ed8ecbfb8f11f6ca05ee75/pandas/core/generic.py#L6655): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

cchwala committed:
minor pep8 fixes

cchwala committed:
fixed parameter order

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -6648,7 +6648,12 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
             * 'outside': Only fill NaNs outside valid values (extrapolate).
 
             .. versionadded:: 0.21.0
+        maxgap : int, optional
```
Probably `max_gap`.

cchwala replied:
Took the name from jreback's suggestion [here](https://github.com/pandas-dev/pandas/issues/12187#issuecomment-221379702). If I am free to chose I would also go with `max_gap`.

jreback replied:
yep ``max_gap`` is good

TomAugspurger commented on pandas/core/missing.py:
```diff
@@ -165,6 +165,16 @@ def interpolate_1d(xvalues, yvalues, method='linear', limit=None,
         raise ValueError('Limit must be greater than 0')
 
+    if (maxgap is not None) and (limit is not None):
+        raise ValueError('maxgap cannot be used together with limit')
```
Is this a fundamental limitation, or just hasn't been implemented yet? I could imagine using both of these at the same time.

If it just hasn't been implemented yet, perhaps raise a NotImplementedError

cchwala replied:
To me, it does not make sense to use both, `maxgap` and `limit`, at the same time, since I want to use `maxgap` to avoid the behavior of `interplotate(limit=N). The purpose of `maxgap` is that NaN-gaps that are too long are not filled in any way. The reasoning is that it can be justified to interpolate, e.g. short periods without sensor data, but it does not make sense to interpolate only some values into a longer gap. 

I could imagine a theoretical use case that somebody would want to interpolate at least some values up to `limit` in a NaN-gap that is longer than `maxgap`. But i do not see why anybody would want to do that with real data.

From the implementation, I think, `limit` and `maxgap` could be combined. We just have to combine their `preserve_nans` indices. I would have to change the logic of the if-statement a little.

jreback commented on pandas/core/generic.py:
```diff
@@ -6648,7 +6648,12 @@ def replace(self, to_replace=None, value=None, inplace=False, limit=None,
+            will be interpolated. For all NaN-gaps wider than that no
+            interpolation is carried out. Must be greater than 0.
 
+            .. versionadded:: 0.25.0
```
need a blank line after

jreback commented on pandas/core/missing.py:
```diff
@@ -182,14 +192,40 @@ def interpolate_1d(xvalues, yvalues, method='linear', limit=None,
-    elif limit_direction == 'backward':
-        preserve_nans = end_nans | set(_interp_limit(invalid, 0, limit))
+    if maxgap is None:
+        if limit_direction == 'forward':
```
can you make this into another function that you call, this is already getting to be a huge function.

jreback commented on pandas/tests/series/test_missing.py:
```diff
@@ -1204,6 +1204,41 @@ def test_interp_limit_to_ends(self):
+        result = s.interpolate(method='linear', maxgap=2, limit_area='inside')
+        assert_series_equal(result, excpected)
+
+        with pytest.raises(ValueError,
```
can you put the fail cases in another test function; i think you need a bit more checking (as you have multiple fail conditions)

jreback reviewed (changes requested):
can you add a couple of examples to the doc-string so its easy to see what this is doing, include using limit and not.

cchwala commented:
@jreback Just a quick comment to let you know that updating this PR is still on my TODO list, but I have not had time to work on it for the last two weeks. Next week looks better, at least judging from now...

jreback commented:
can you merge master

cchwala committed:
Merge remote-tracking branch 'upstream/master' into interpolate_maxgap

cchwala commented:
Resolved the tiny merge conflict. Still no other progress with the PR. Sorry. I am too overloaded till early April.

jreback commented:
@cchwala if you have time; this is a nice patch. pls update to comments.

jreback commented:
@pandas-dev/pandas-core if someone wants to take this over the line, pls merge master and update to comments.

cchwala committed:
Merge remote-tracking branch 'upstream/master' into interpolate_maxgap

cchwala commented:
@jreback Sorry for the long silence. I will do it today.

cchwala committed:
Changed parameter name from `maxgap` to `max_gap`

cchwala committed:
Moved code to derive indices of "NaNs to preserve" in separate function

cchwala committed:
Tests for errors extended and moved to own function

cchwala committed:
added blank lines in docstring as requested

cchwala committed:
Added test which fails for method='pad'

For method='pad' the `max_gap` keyword does not seem to have an effect.

cchwala commented:
There is a problem. For `s.interpolate(method='pad', max_gap=2)` the `max_gap` keyword does not seem to have an effect. I added a test which fails in this case.

From a first quick search the cause might lie here:
https://github.com/pandas-dev/pandas/blob/b8ad9da4213be55c587db015c640d927aeb7feaf/pandas/core/internals/blocks.py#L1096-L1116
There are two different pathways for `interpolate` depending on the selected `method`. For `pad`, `ffill` and `bfill` `missing.interpolate_2d` is used which does not yet support the `max_gap` option. It also does not seem to recognize the keywords `limit_direction` and `limit_area`, they are silently ignored.

cchwala committed:
Merge remote-tracking branch 'upstream/master' into interpolate_maxgap

cchwala committed:
manually add black code formating

cchwala commented:
I resolved the merge conflict. However, I still have not found the time to work further on this PR, in particular because it somehow is blocked by #26796. It remains somewhere in the middle of my TODO list...

cchwala committed:
First WIP but working version to fix issue with `pad` and `limit_area`

* Test are green for test_missing.py but some red test appear
  when running the whole test suite...
* The current solution uses `missing._derive_indices_of_nans_to_preserve`,
  which was implemented some commits ago also when using `pad` and
  `backfill`.
* The implementation is not very elegant (yet) since I introduced
  a second version of `missing.interpolate_2d` only for the 1D-case of
  using `pad` and `backfill`, called `missing.interpolate_1d_fill`.
  Maybe distinguishing 1D and 2D could also be done in the original
  `missing.interpolate_2d`...

cchwala committed:
fix: do not decide based on dimension but on crucial kwargs which interpolation function to use

cchwala committed:
some clean up

* added explanatory docstring to missing.interpolate_1d_fill
* ran black code formating (and shortend test data code to fit
  onto one line)

cchwala commented:
I have introduced a solution for #26796 so that the implementation of the new `max_gap` kwarg also works as expected when using `method=pad` . I put these changes all in this PR since they depend on the function `missing._derive_indices_of_nans_to_preserve`, which was introduced earlier in this PR.

The code now distinguishs between the 1D- and 2D-case of doing interpolation with `pad` or `backfill`. In case of 1D, the approach is similar to the one using the scipy-interpolate-machinery.

The code will need some polishing, but before I continue I want to know if the implementation is acceptable in general.

Questions to @jreback @TomAugspurger or others that are familiar with the interpolation code:
1. Should I continue with this approach?
1. Should I simplify the new [`missing.interpolate_1d_fill()`](https://github.com/pandas-dev/pandas/pull/25141/files/c72acdb904cc2875bd530a1b780a809082ab16bb..12d2e5b1408f058fa4eb09082e1472a58ecf4818#diff-0cca2dbee24098c1488ed79b05b7248fR521) to use [`pad_1d` and `backfill_1d`](https://github.com/pandas-dev/pandas/blob/12d2e5b1408f058fa4eb09082e1472a58ecf4818/pandas/core/missing.py#L679:L688) instead of the current approach using the [2D-code
](https://github.com/pandas-dev/pandas/blob/12d2e5b1408f058fa4eb09082e1472a58ecf4818/pandas/core/missing.py#L577:L582) that I copied from the original [`missing.interpolate_2d`](https://github.com/pandas-dev/pandas/blob/12d2e5b1408f058fa4eb09082e1472a58ecf4818/pandas/core/missing.py#L601)

TomAugspurger commented:
I don't have a strong prior on whether this is a good approach, but your description sounds reasonable.

cchwala committed:
Make it work with NaT and test for that

cchwala committed:
Added comment on why two interpolate fill functions are needed

cchwala committed:
fix typo

cchwala committed:
Added tests for DataFrames

The one with `method=pad` currently fails. The others pass.

cchwala commented:
Digging further into this and adding more test for `method=pad` I ran into https://github.com/pandas-dev/pandas/issues/12918

Looking for a fix now...

cchwala committed:
Added failing test for https://github.com/pandas-dev/pandas/issues/12918

cchwala committed:
Now using 1D pad and backfill functions in `interpolate_1d_fill()`

Also added test that now passes with using the 1D pad functions

cchwala commented:
Note to myself (and everybody interested):

There might be an inconsistency with how axes are determined in relation to the actual data stored for a DataFrame in `._data`, which might have reversed axes, indicated by `DataFrame._AXIS_REVERSED`.

[`NDFrame.interpolate()` has some logic to set the `axis`](https://github.com/pandas-dev/pandas/blob/367670e80ff008d77c791437a7a5df3e1abe012b/pandas/core/generic.py#L7001:L7009) while ['NDFrame.fillna()`](https://github.com/pandas-dev/pandas/blob/367670e80ff008d77c791437a7a5df3e1abe012b/pandas/core/generic.py#L6232:L6233) just sets a default. Both call `NDFrame._data.interpolate()` then using the axis. 

When using `DataFrame.pad()` vs `DataFrame.interpolate(method=`pad`) the `axis` in `NDFrame._data.interpolate()` are different. Unfortunately some simple fixes that I tried made several other test fail. So no fix for now...

cchwala commented:
FYI, I am offline till 7th of October but plan to continue with this PR afterwards

WillAyd commented:
@cchwala is this still active?

cchwala committed:
Merge remote-tracking branch 'upstream/master' into interpolate_maxgap

# Conflicts:
#	pandas/core/internals/blocks.py
#	pandas/core/missing.py

cchwala commented:
@WillAyd Not really active... But I resolved the merge conflicts on my machine.

Unfortunately I am struggling with #29330, even in a fresh miniconda3 install, so I cannot run the test.

cchwala committed:
Additional required adjustments after merge with upstream/master

cchwala committed:
Merge remote-tracking branch 'upstream/master' into interpolate_maxgap

cchwala committed:
Removed test for bug with pad which should be solved in a separate PR

cchwala committed:
removed trailing whitespaces

cchwala committed:
fixed formating for black and flake8

cchwala committed:
updated docstring for interpolat with max_gap

* added example
* optimized existing text

cchwala committed:
added max_gap info and example to documentation

cchwala committed:
added info to whatsnew file

cchwala committed:
flake8

cchwala committed:
update docs with info on limit_direction and method pad

cchwala committed:
better test for https://github.com/pandas-dev/pandas/issues/26796

cchwala committed:
typo, black, flake8

cchwala committed:
update to doc

cchwala committed:
fix wrong behavior when combining max_gap and limit_direction

* limit_direction was not considered before when max_gap was provided
* test have been adjusted for the new correct behavior and additional ones have been added

cchwala commented:
@TomAugspurger @WillAyd This is now finally ready for another review.

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -7085,6 +7125,28 @@ def interpolate(
+        if (method == "pad") or (method == "ffill"):
+            if (limit_direction == "backward") or (limit_direction == "both"):
+                raise ValueError(
+                    "`limit_direction` must not be `%s` for method `%s`"
```
You can use f-strings for this and the one on L 7140

cchwala replied:
okay

TomAugspurger commented on pandas/core/missing.py:
```diff
@@ -306,6 +293,82 @@ def interpolate_1d(
+
+    if max_gap is not None:
+
+        def bfill_nan(arr):
```
What's the benefit to making this a separate closure?

cchwala replied:
There is no real reason. Maybe at some point I thought the function definition would make things clearer.

Should I just put the content of the function in-line starting at L350?

TomAugspurger commented on pandas/core/missing.py:
```diff
@@ -471,6 +534,74 @@ def _akima_interpolate(xi, yi, x, der=0, axis=0):
+        # convert float back to datetime64
+        values = values.astype(orig_values.dtype)
+
+    # if np.issubdtype(values.dtype, np.datetime64):
```
Any reason this is commented out?

cchwala replied:
I just forgot to remove it. This stems from when I tried to manually chose the correct fill_value. But getting the correct fill_value is handled [here](https://github.com/pandas-dev/pandas/pull/25141/files#diff-8f51ce8697bdff680b3f71b196c582eeR1176)

TomAugspurger reviewed (commented):
There's a lot going on here, sorry only got through part of it at a glance.

Is there any chance that the changes to allow `limit_area` and `limit_direction` in Series.interpolate with `pad` can be split into its own PR? Would that make this one much smaller?

cchwala commented:
> Is there any chance that the changes to allow `limit_area` and `limit_direction` in Series.interpolate with `pad` can be split into its own PR? Would that make this one much smaller?

For making `max_gap` work with `pad` I had to introduce [missing.interpolate_1d_fill](https://github.com/pandas-dev/pandas/pull/25141/files#diff-0cca2dbee24098c1488ed79b05b7248fR537) because the existing`missing.interpolate_2d`, which `pad` did always use, is not (easily) extendable. The cause for the wrong behavior of `limit_area` when using `pad` was the same, so this was solved somehow as a byproduct.

If we would want to split this PR up, probably most changes of this PR would move to the one to solve the issue with `pad` and `limit_area`. So yes, this one would be much smaller, but the other one would be equally large.

WillAyd commented:
@cchwala can you merge master and fix up the CI error for code checks?

cchwala commented:
@WillAyd I separated large parts of the changes into #31048 and will continue here afterwards.

WillAyd commented:
Closing to clean queue - I think can revisit after #31048
